pkg/controller/iam/managed_test.go

```diff
@@ -0,0 +1,797 @@
+	uniqueID    = fqName
+)
+
+type strange struct {
+	resource.Managed
+}
+
+type valueModifier func(*v1beta1.ServiceAccount)
+
+func withName(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Status.AtProvider.Name = s }
+}
+
+func withProjectID(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Status.AtProvider.ProjectID = s }
+}
+
+func withDisplayName(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Spec.ForProvider.DisplayName = &s }
+}
+
+func withDescription(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Spec.ForProvider.Description = &s }
+}
+
+func withUniqueID(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Status.AtProvider.UniqueID = s }
+}
+
+func withEmail(s string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Status.AtProvider.Email = s }
+}
+
+func withDisabled(b bool) valueModifier {
+	return func(i *v1beta1.ServiceAccount) { i.Status.AtProvider.Disabled = b }
+}
+
+func withExternalNameAnnotation(externalName string) valueModifier {
+	return func(i *v1beta1.ServiceAccount) {
+		if i.ObjectMeta.Annotations == nil {
+			i.ObjectMeta.Annotations = make(map[string]string)
+		}
+		i.ObjectMeta.Annotations[wtfConst] = externalName
+	}
+}
+
+func serviceAccount(im ...valueModifier) *v1beta1.ServiceAccount {
+	sa := &v1beta1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:       metadataName,
+			Finalizers: []string{},
+		},
+		Spec: v1beta1.ServiceAccountSpec{
+			ResourceSpec: runtimev1alpha1.ResourceSpec{
+				ProviderReference: &corev1.ObjectReference{Name: providerName},
+				WriteConnectionSecretToReference: &runtimev1alpha1.SecretReference{
+					Namespace: namespace,
+					Name:      connectionSecretName,
+				},
+			},
+			ForProvider: v1beta1.ServiceAccountParameters{
+				DisplayName: &displayName,
+			},
+		},
+	}
+
+	for _, m := range im {
+		m(sa)
+	}
+
+	return sa
+}
+
+var _ managed.ExternalClient = &external{}
+var _ managed.ExternalConnecter = &connecter{}
+
+func TestConnect(t *testing.T) {
+	provider := gcpv1alpha3.Provider{
+		ObjectMeta: metav1.ObjectMeta{Name: providerName},
+		Spec: gcpv1alpha3.ProviderSpec{
+			ProjectID: project,
+			ProviderSpec: runtimev1alpha1.ProviderSpec{
+				CredentialsSecretRef: &runtimev1alpha1.SecretKeySelector{
+					SecretReference: runtimev1alpha1.SecretReference{
+						Namespace: namespace,
+						Name:      providerSecretName,
+					},
+					Key: providerSecretKey,
+				},
+			},
+		},
+	}
+
+	secret := corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: providerSecretName},
+		Data:       map[string][]byte{providerSecretKey: []byte(providerSecretData)},
+	}
+
+	type strange struct {
+		resource.Managed
+	}
+
+	type args struct {
+		ctx context.Context
+		mg  resource.Managed
+	}
+	type want struct {
+		err error
+	}
+
+	cases := map[string]struct {
```

Comment:
I've been trying to get into the habit of including a human-readable `reason string` with each test case that's printed when it fails in order to help document their intent a little better. Could be overkill here though, as the subtest names are fairly self-explanatory to me.